sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisReader.java

```diff
@@ -84,6 +85,11 @@ public boolean advance() throws IOException {
       for (int i = 0; i < shardIterators.size(); ++i) {
         currentRecord = shardIterators.getCurrent().next();
         if (currentRecord.isPresent()) {
+          Instant approximateArrivalTimestamp = currentRecord.get()
+              .getApproximateArrivalTimestamp();
+          if (approximateArrivalTimestamp.isAfter(watermark)) {
+            watermark = approximateArrivalTimestamp;
```

Comment:
This doesn't work nicely with a scheme where we read data from shards in round robin way, because you may advance watermark based on a first shared although you've got older data in other shards.

What about changing this scheme to actually read the shard where records are most outdated? This will improve the whole solution greatly.

Reply:
I know it's not ideal but I would treat it as a temporary solution. I'm planning to change the way the streams are read in [BEAM-2468](https://issues.apache.org/jira/browse/BEAM-2468)

Reply:
Oh interesting, I didn't realize that the number of shards is fixed. Then at some point this connector can be considerably simplified using Splittable DoFn, by decomposing it into two DoFn's - one (non-splittable) listing the shards, and the other (splittable) reading a single shard. See https://beam.apache.org/blog/2017/08/16/splittable-do-fn.html .

Reply:
The number of shards is not fixed actually but the stream resharding is not supported currently by this connector. I'm planning to fix that in https://issues.apache.org/jira/browse/BEAM-2469.
I've heard about this Splittable DoFns idea some time ago but didn't delve into details yet. Sounds really interesting!